functions/index.js

```diff
@@ -1,95 +1,176 @@
+const textModel = 'gemini-1.5-flash';
+// const visionModel = 'gemini-1.0-pro-vision';
+async function saveNewLanguageCode(languagesCollection, detectedLanguage, languages) {
+  if (detectedLanguage && !languages.includes(detectedLanguage)) {
+    console.log(`Adding new language code: ${detectedLanguage}`);
+    try {
+      await languagesCollection.add({ code: detectedLanguage });
+      console.log("Successfully added new language code to database");
+      return true;
+    } catch (error) {
+      console.error("Error adding new language code to database:", error);
+      return false;
+    }
+  }
+  return false;
+}
+const vertexAI = new vertexAIApi.VertexAI({ project: project, location: location });
 
-const generativeVisionModel = vertexAI.getGenerativeModel({
-    model: visionModel,
-});
+// const generativeVisionModel = vertexAI.getGenerativeModel({
+//     model: visionModel,
+// });
 
 const generativeModelPreview = vertexAI.preview.getGenerativeModel({
-    model: textModel,
+  model: textModel,
 });
 
-const generationConfig = {
+// use onDocumentWritten here to prepare to "edit message" feature later
+exports.onChatWritten = v2.firestore.onDocumentWritten("/public/{messageId}", async (event) => {
+
+  const document = event.data.after.data();
+  const message = document["message"];
+  console.log(`message: ${message}`);
+
+  // no message? do nothing
+  if (message == undefined) {
+    return;
+  }
+  const curTranslated = document["translated"];
+
+  // check if message is translated
+  if (curTranslated != undefined) {
+    // message is translated before, 
+    // check the original message
+    const original = curTranslated["original"];
+
+    console.log('Original: ', original);
+    // message is same as original, meaning it's already translated. Do nothing
+    if (message == original) {
+      return;
+    }
+  }
+    // Get list of languages from Firestore
+    const db = admin.firestore();
+    const languagesCollection = db.collection("languages");
+    const languagesSnapshot = await languagesCollection.get();
+    const languages = languagesSnapshot.docs.map((e) => e.data().code);
+    console.log("Current languages in database:", languages);
+
+  const generationConfig = {
     temperature: 1,
     topP: 0.95,
     topK: 64,
     maxOutputTokens: 8192,
     responseMimeType: "application/json",
     responseSchema: {
       type: "object",
-      properties: {
-        en: {
-          type: "string"
-        }
+      properties:
+      {
+        "detectedLanguage": {
+          type: "string",
+          description: "The ISO 639-1 language code that was detected",
+          pattern: "^[a-z]{2}$"
+        },
+        "translation": languages.reduce((acc, lang) => {
+          acc[lang] = { type: "string" };
+          return acc;
+        }, {}),
       },
-      required: [
-        "en"
-      ]
+      required: ["detectedLanguage", "translation"]
```

Comment:
i tried to run it with message in vietnamese, and the detectedLanguage is returned as 'und',not sure what language code is that